### DISCUSSION TITLE
Improve Run/Debug Tests

### DISCUSSION THREAD

**Discussion by Structed** [MikeSchulze/gdUnit4#138] [Ideas]
As I was getting started with GDUnit4, it took me a while to figure out that I cannot just "Run All Tests" from the GDUnit Panel in the Editor (See screenshot). What I later discovered as a "Rerun tests" looked like "Run all Tests" to me!

![image](https://user-images.githubusercontent.com/29922/223220147-428b4d07-7714-4d06-b06a-50cf9926237a.png)

I'd like to propose a few things:

1. Expand documentation how to run tests:
  Detail all the places you can run/debug tests from - which is super flexible!
  For example, that you can actually right-click on individual files & folders in the _FileSystem_!
2. Document that, wherever you choose to run tests, this action defines the subset of tests to be run and therefore to be re-run when you click the "Re-run" button in the GDUnit panel.
3. Introduce hotkeys to:
    * Run/Debug all tests
    * Re-Run/Debug previous tests

What do you think, @MikeSchulze?
I'd be happy to contribute documentation, if you like the proposal.

**Comment by MikeSchulze** [comment]
1) This is already documented, 
https://mikeschulze.github.io/gdUnit3/first_steps/firstTest/
https://mikeschulze.github.io/gdUnit3/faq/inspector/
But yes, a short overview of available options can be documented in a separate page for better overview
2) These buttons have a tooltip and show what they are for ;)
![image](https://user-images.githubusercontent.com/347037/223357585-0dd6e025-0f12-4fd0-af2d-036714c59a63.png)
3) Yes we should add hot keys to trigger this actions

**Reply by Structed** [reply]
haha, yes, I eventually saw the tooltips, but coming from other IDEs (JetBrains Rider, Visual Studio), the first button is ALWAYS run all tests. I think there's an opportunity to do the same and add a run all tests button to the toolbar?
![image](https://user-images.githubusercontent.com/29922/223388676-134e006b-6088-4b07-815e-ada67270822a.png)
![image](https://user-images.githubusercontent.com/29922/223388793-132426a4-23f5-4da2-aeb3-547ab2077a6c.png)


Yeah, I was aware of the right-click in editor, but not the other options. Hence I was asking to expand the docs.

Also, the [Inspector FAQ](https://mikeschulze.github.io/gdUnit3/faq/inspector/) explicitly says "_(Re)run Tests_" (yes, nit-picking 😛).

In the end, I just wanted to make a point that I found it non-intuitive to fun all tests and it took me quite some time to figure it out, as Run-all is my default workflow.
Specifically, when I add a test, I just want to run all including the new one.

**Comment by MikeSchulze** [comment]
Adding a "Run all tests" button is not a problem ;)
Just to be sure you mean all the tests in the current test suite and not all tests overall


> Also, the [Inspector FAQ](https://mikeschulze.github.io/gdUnit3/faq/inspector/) explicitly says "(Re)run Tests" (yes, nit-picking 😛).

(Re)run means not run all ;)
It means rerun the last executed test's.

> Specifically, when I add a test, I just want to run all including the new one.
* to run a single test case by select the test name and run
* to run all tests of a test suite by click somewer in your test suite and run
* to run a set of test suites just select it in the file navigator and run
* to run all test suite select the test root folder and run


I will improve the documentation to be more clear here, thanks for your input.
Adding a extra button to run all tests of the active test suite [GD-140](https://github.com/MikeSchulze/gdUnit4/issues/140)

**Reply by Structed** [reply]
Thanks so much! 🙏
I actually meant all tests in the project, not only the suite! That's also the experience you get from other IDEs.

**Reply by MikeSchulze** [reply]
> Thanks so much! 🙏 I actually meant all tests in the project, not only the suite! That's also the experience you get from other IDEs.

Ah ok, run all tests by default is a very bad option, it can takes a very long time according to your amount of tests.
Run all tests overall is typicaly a CI task and should be never the default execution.

You can do this just by
* to run all test suite select the test root folder and run

But I can also add this kind of button, but it will be hidden by default and must be enabled extra under GdUnit4 properties.

**Reply by Structed** [reply]
Well, depends: if you have a small add-on with fast test (like me), I think it totally makes sense to run all! 😊 I really like the idea to add a button which is hidden by default. Good compromise - thank you!
In fact, a hotkey for that would also be cool. I would rarely use the button anyway if there's a hotkey.

**Comment by Structed** [comment]
Implemented in #140 - thank you very much!

**Comment by MikeSchulze** [comment]
With GdUnit4 v4.1.0 the new run overall button is now optional and can be configured under GdUnit settings.
![image](https://user-images.githubusercontent.com/347037/227868775-d2515d3e-1ad2-41ec-8a76-7f8b3ff17da3.png)
You also allowed to use shortcuts to run/debug/create tests and can be configured under GdUnit settings too.
![image](https://user-images.githubusercontent.com/347037/227868999-ff1209ec-89f8-4b0a-8b34-a06c80f54a96.png)

**Reply by Structed** [reply]
Thank you!
